File: 240314/십자 모양 폭발/cross-shape-bomb.py
n=int(input())
arr=[list(map(int,input().split())) for _ in range(n)]
temp=[[0]*n for _ in range(n)]

row,col=map(int,input().split())

row=row-1
col=col-1

power=arr[row][col]

for i in range(power):

    if row+i<n:
        arr[row+i][col]=0
    if col+i<n:
        arr[row][col+i]=0
    if row-i>=0:
        arr[row-i][col]=0
    if col-i >=0:
        arr[row][col-i]=0



for col in range(n):
    temp_start_row=n-1
    for row in range(n-1,-1,-1):
        if arr[row][col]!=0:
            temp[temp_start_row][col]=arr[row][col]
            temp_start_row-=1
        

for i in range(n):
    for j in range(n):
        print(temp[i][j],end=' ')
    print()

File: 250327/십자 모양 폭발/cross-shape-bomb.java
import java.util.Scanner;

public class Main {
    static int n;
    static int row;
    static int col;
    static int[][] arr;
    static int[][] temp;

    static void boom() {
        int size = arr[row][col];
        arr[row][col] = 0;
        int x = row;
        int y = col;
        for (int i = 1; i < size; i++) {
            if (x - 1 >= 0) {
                arr[--x][col] = 0;
            }
        }

        for (int i = 1; i < size; i++) {
            if (y + 1 < n) {
                arr[row][++y] = 0;
            }
        }
        x = row;
        y = col;
        for (int i = 1; i < size; i++) {
            if (x + 1 < n) {
                arr[++x][col] = 0;
            }
        }

        for (int i = 1; i < size; i++) {
            if (y - 1 >= 0) {
                arr[row][--y] = 0;
            }
        }

    }

    static void gravity() {
        for(int i=0; i<n;i++){
            gravityCell(i);
        }
    }
    static void gravityCell(int col){
        int[] temp=new int[n];
        int idx=0;
        for(int i=0; i<n;i++){
            if(arr[n-1-i][col]==0) continue;
            temp[idx++]=arr[n-1-i][col];
        }
        
        for(int i=0; i<n;i++){
            arr[n-1-i][col]=temp[i];
        }
               
    }
    


    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        arr = new int[n][n];
        temp = new int[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                arr[i][j] = sc.nextInt();
            }
        }

        row = sc.nextInt()-1;
        col = sc.nextInt()-1;

        boom();
        gravity();
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                System.out.print(arr[i][j]+" ");
            }
            System.out.println();
        }

        System.out.println();


    }
}
